README
This is a fork of CCZE, a fast log colorizer written in C, intended to be a
drop-in replacement for colorize (http://colorize.raszi.hu).

See the NEWS file for a description of recent changes to CCZE.

See the file INSTALL for instructions on how to build and install the
CCZE data and program files. See the CCZE manual for details about
configuration. Type "man ccze" in the shell prompt.

Please visit the official CCZE web page for more information. The URL
is http://bonehunter.rulez.org/CCZE.html.

====================================================================================
This fork adds a plugin to permit the colorizing of DXSpider logs.
At present, it colorizes the chanerr debug output. 

To use it with your DXSpider server, open a DXSpider console and enter:

unset/debug nologchan
set/debug chanerr

Then, log in as a normal user on a virtual consol or open an xterm window and
enter the following:

watchdbg -2 chanerr | ccze

ccze will recognize the DXSpider log file format and colorize the output from watchdbg.
If you are using a slow computer, your can slightly reduce ccze cpu workload by specifying
the appropriate ccze plugin. There are two DXSpider plugins and you will generally want
run with them both loaded: DXSpider and DXSpiderPCxx. Enter:

watchdbg -2 chanerr | ccze --plugin=DXSpider,DXSpiderPCxx

Note, As of March 2023, the watchdbg application provided with the mojo version of DXSpider
buffers its output. Thus, if you wish to pipe its output to ccze in real time, you will need
to make a one line change to /spider/perl/watchdbg. Near line 52, add the command

STDOUT->autoflush(1);

to automatically flush the standard output buffer:


-------------------------------- perl/watchdbg --------------------------------
index 2b1986a7..02084284 100755
@@ -52,6 +52,7 @@ while (@ARGV) {
 
 
 # seek to end of file
+STDOUT->autoflush(1);
 $fh->seek(0, 2);
 for (;;) {
 	my $line = $fh->getline;



Changelog
commit b345059da45d99bde54ca2f21454eeb5654f27a6
Author: John Kuras <w7og@yahoo.com>
Date:   Tue Jun 13 18:51:44 2023 -0700

    code cleanup

M	src/mod_DXSpider.c

commit a63cf1cbc4ebf6d188391586332157a6daf54004
Author: John Kuras <w7og@yahoo.com>
Date:   Tue Jun 13 18:46:09 2023 -0700

    added DXSpider test to the test suite

M	testsuite/Makefile.in
A	testsuite/bug-dxspider.in
A	testsuite/bug-dxspider.ok
A	testsuite/bug-dxspider.test

commit 8601a116f55338ab695217def5a9b01a41b5698e
Merge: 62729f2 93af421
Author: John Kuras <w7og@yahoo.com>
Date:   Tue Jun 13 03:08:32 2023 -0700

    Merge remote-tracking branch 'refs/remotes/origin/master'

commit 62729f2eccf693d1d4c32e0a82416c8d4c9cc01c
Author: John Kuras <w7og@yahoo.com>
Date:   Tue Jun 13 03:05:51 2023 -0700

    created ccze color items for use in the DXSpider modules

M	src/ccze-color.c
M	src/ccze.h
M	src/mod_DXSpider.c
M	src/mod_DXSpiderPCxx.c

commit 9b0e70089b755228bb09031136381a44117651d6
Author: John Kuras <w7og@yahoo.com>
Date:   Tue Jun 13 03:04:11 2023 -0700

    updated gitignore

M	.gitignore

commit d4bd2eae4dd280e1582dcc05f69cd3d0d14ba1c4
Author: John Kuras <w7og@yahoo.com>
Date:   Tue Jun 13 03:01:42 2023 -0700

    added man page for config file

M	Rules.mk.in
M	doc/Makefile.in
A	doc/ccze-config.4.in
M	doc/ccze.1.in

commit 4752810a306bd900c2840d5f2af52b8e70f1a704
Author: John Kuras <w7og@yahoo.com>
Date:   Tue Jun 13 02:58:18 2023 -0700

    updated AUTHORS

M	AUTHORS

commit 93af4212214c0dbed08a4c69d8ab2912d4042f29
Author: dad <w7og@yahoo.com>
Date:   Sun Jun 11 23:34:22 2023 -0700

    Update README

M	README

commit 73175f8b4a6757ecdadaad90f3eaefba35d7c65b
Author: John Kuras <w7og@yahoo.com>
Date:   Sun Jun 11 22:43:02 2023 -0700

    added colorizer plugin got DXSpider PCxx messages

M	ccze.plugins
M	src/mod_DXSpider.c
A	src/mod_DXSpiderPCxx.c

commit 302feff3b806e56d69301f0f1bc94d4335a6e04f
Author: root <root@kuras-sea.mylocal>
Date:   Sun Jun 11 00:33:05 2023 -0700

    added a module for DXSpider debug log files

M	ccze.plugins
A	src/mod_DXSpider.c

commit e00d1db1a08d42ee3f9385302fea708af16c0695
Author: root <root@kuras-sea.mylocal>
Date:   Sun Jun 11 00:31:51 2023 -0700

    updated version number

M	Rules.mk.in
M	configure.ac

commit e01d470c71cea39b009f03b8e4e514703cb6dffa
Author: root <root@kuras-sea.mylocal>
Date:   Sun Jun 11 00:30:23 2023 -0700

    updated gitignore

M	.gitignore

commit 7c7927fba98275c10ae614bd8c1a1367f87a258c
Author: Gergely Nagy <algernon@madhouse-project.org>
Date:   Thu Aug 28 12:04:32 2014 +0200

    src/ccze.c: Default to raw ANSI if stdout is not a terminal
    
    If stdout is not a terminal, use raw ANSI by default, as pagers handle
    that far better than curse. This fixes #1.
    
    Reported-by: Marcel Partap
    Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

M	src/ccze.c

commit b81e8be6f2d5f404fe9fee1adc1d1dcee652da09
Author: Gergely Nagy <algernon@madhouse-project.org>
Date:   Thu Jun 26 16:49:03 2014 +0200

    Upgrade config.* and install-sh
    
    Update the config.guess, config.sub and install-sh files from a recent
    automake. This will all go away if and when ccze is switched to
    automake.
    
    Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

M	config.guess
M	config.sub
M	install-sh

commit efcfd8dcb85239100c4f7ca181cf0400377efc67
Author: Gergely Nagy <algernon@madhouse-project.org>
Date:   Thu Jun 26 16:22:16 2014 +0200

    debian/: Drop Debian packaging too
    
    Packaging does not belong to the upstream master branch, drop it.
    
    Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

D	debian/changelog
D	debian/control
D	debian/copyright
D	debian/rules

commit e116ae590a90aed294420ecfac3421cdf163b017
Author: Gergely Nagy <algernon@madhouse-project.org>
Date:   Thu Jun 26 16:21:31 2014 +0200

    Drop the spec file
    
    Packaging is not something upstream does - drop the spec file.
    
    Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

M	.gitignore
M	Makefile.in
D	ccze.spec.in

commit b5911afb6991572a0652540b0191a6a53b8fc339
Author: Gergely Nagy <algernon@madhouse-project.org>
Date:   Thu Jun 26 16:17:16 2014 +0200

    Drop the old TLA ChangeLogs
    
    Drop the old TLA ChangeLogs, they're not being updated anymore since the
    switch to Git. While there, also updated all users of PATCHLEVEL to use
    VERSION (defined in Rules.mk.in).
    
    Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

D	ChangeLog
D	ChangeLog-0.1
M	Makefile.in
M	Rules.mk.in
M	ccze.spec.in
M	debian/rules
M	doc/Makefile.in
M	src/Makefile.in
M	src/ccze-cssdump.c
M	src/ccze-dump.c
M	src/ccze.c

commit a1d682fd7f97bb16b10a5ca4962dbbed9e9cc8cf
Author: Gergely Nagy <algernon@madhouse-project.org>
Date:   Thu Jun 26 16:00:06 2014 +0200

    src/ccze.c: Minor exit-time leak plugs
    
    Free up a few more things at exit-time to make valgrind 100% happy.
    
    Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

M	src/ccze.c

commit 1fb3f9ec1e39b26e2831e8d1045fe38bbf76bc4e
Author: Gergely Nagy <algernon@madhouse-project.org>
Date:   Thu Jun 26 15:53:24 2014 +0200

    src/ccze-color.c: Use a static stat buffer
    
    Use a static struct stat buffer when loading a color file, so we have a
    clean starting state.
    
    Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

M	src/ccze-color.c

commit 1de7a6e285df3aabdfd6870b6c08b0645ebd9523
Author: Gergely Nagy <algernon@madhouse-project.org>
Date:   Thu Jun 26 15:41:18 2014 +0200

    Simplify the compiler warning flags
    
    Instead of having a huge list of custom compiler warning flags, use
    -Wall -Wextra only.
    
    Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

M	configure.ac
M	src/Makefile.in

commit ef49cf13fbb6883cb77bab9e25f70d66e9d7f927
Author: Gergely Nagy <algernon@madhouse-project.org>
Date:   Thu Jun 26 15:38:53 2014 +0200

    Warning fixes
    

src/ccze.h
/* -*- mode: c; c-file-style: "gnu" -*-
 * ccze.h -- Public interface to CCZE.
 * Copyright (C) 2002, 2003 Gergely Nagy <algernon@bonehunter.rulez.org>
 *
 * This file is part of ccze.
 *
 * ccze is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * ccze is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 * License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#ifndef _CCZE_H
#define _CCZE_H 1

#ifdef HAVE_SYSTEM_H
# include "system.h"
#endif
#include <pcre.h>
#include <ncurses.h>
#include <stddef.h>

/* Compatibility */
void *ccze_malloc (size_t size);
void *ccze_realloc (void *ptr, size_t size);
void *ccze_calloc (size_t nmemb, size_t size);

/* Colors */
typedef enum
{
  CCZE_COLOR_DATE = 0,
  CCZE_COLOR_HOST,
  CCZE_COLOR_PROC,
  CCZE_COLOR_PID,
  CCZE_COLOR_PIDB,
  CCZE_COLOR_DEFAULT,
  CCZE_COLOR_EMAIL,
  CCZE_COLOR_SUBJECT,
  CCZE_COLOR_DIR,
  CCZE_COLOR_FILE,
  CCZE_COLOR_SIZE,
  CCZE_COLOR_USER,
  CCZE_COLOR_HTTPCODES,
  CCZE_COLOR_GETSIZE,
  CCZE_COLOR_HTTP_GET,
  CCZE_COLOR_HTTP_POST,
  CCZE_COLOR_HTTP_HEAD,
  CCZE_COLOR_HTTP_PUT,
  CCZE_COLOR_HTTP_CONNECT,
  CCZE_COLOR_HTTP_TRACE,
  CCZE_COLOR_UNKNOWN,
  CCZE_COLOR_GETTIME,
  CCZE_COLOR_URI,
  CCZE_COLOR_IDENT,
  CCZE_COLOR_CTYPE,
  CCZE_COLOR_ERROR,
  CCZE_COLOR_PROXY_MISS,
  CCZE_COLOR_PROXY_HIT,
  CCZE_COLOR_PROXY_DENIED,
  CCZE_COLOR_PROXY_REFRESH,
  CCZE_COLOR_PROXY_SWAPFAIL,
  CCZE_COLOR_DEBUG,
  CCZE_COLOR_WARNING,
  CCZE_COLOR_PROXY_DIRECT,
  CCZE_COLOR_PROXY_PARENT,
  CCZE_COLOR_SWAPNUM,
  CCZE_COLOR_PROXY_CREATE,
  CCZE_COLOR_PROXY_SWAPIN,
  CCZE_COLOR_PROXY_SWAPOUT,
  CCZE_COLOR_PROXY_RELEASE,
  CCZE_COLOR_MAC,
  CCZE_COLOR_VERSION,
  CCZE_COLOR_ADDRESS,
  CCZE_COLOR_NUMBERS,
  CCZE_COLOR_SIGNAL,
  CCZE_COLOR_SERVICE,
  CCZE_COLOR_PROT,
  CCZE_COLOR_BADWORD,
  CCZE_COLOR_GOODWORD,
  CCZE_COLOR_SYSTEMWORD,
  CCZE_COLOR_INCOMING,
  CCZE_COLOR_OUTGOING,
  CCZE_COLOR_UNIQN,
  CCZE_COLOR_REPEAT,
  CCZE_COLOR_FIELD,
  CCZE_COLOR_CHAIN,
  CCZE_COLOR_PERCENTAGE,
  CCZE_COLOR_FTPCODES,
  CCZE_COLOR_KEYWORD,
  CCZE_COLOR_PKGSTATUS,
  CCZE_COLOR_PKG,
  CCZE_COLOR_DXSCHAN,
  CCZE_COLOR_DXSCHANDIRLEFT,
  CCZE_COLOR_DXSCHANDIRRIGHT,
  CCZE_COLOR_DXSCHANDIRX,
  CCZE_COLOR_DXSCHANNAME,
  CCZE_COLOR_DXSCHANERROR, 
  CCZE_COLOR_DXSCHANERRORTYPE,
  CCZE_COLOR_DXSPROGRESS,
  CCZE_COLOR_DXSPROGRESSTYPE,
  CCZE_COLOR_DXSDXCALL,
  CCZE_COLOR_DXSDXFREQUENCY,
  CCZE_COLOR_DXSDXSPOTTER,
  CCZE_COLOR_DXSNODE,
  CCZE_COLOR_DXSDXCOMMENT,
  CCZE_COLOR_DXSDXROUTE,
  CCZE_COLOR_DXSPC9XNUM,
  CCZE_COLOR_DXSPCNUM,
  CCZE_COLOR_DXSPCDELIM,
  CCZE_COLOR_DXSPCMESSAGENUMS,
  CCZE_COLOR_DXSPCMESSAGEWORDS,
  CCZE_COLOR_DXSPCHOPS,  

  CCZE_COLOR_STATIC_BLACK,
  CCZE_COLOR_STATIC_RED,
  CCZE_COLOR_STATIC_GREEN,
  CCZE_COLOR_STATIC_YELLOW,
  CCZE_COLOR_STATIC_BLUE,
  CCZE_COLOR_STATIC_CYAN,
  CCZE_COLOR_STATIC_MAGENTA,
  CCZE_COLOR_STATIC_WHITE,
  CCZE_COLOR_STATIC_BOLD_BLACK,
  CCZE_COLOR_STATIC_BOLD_RED,
  CCZE_COLOR_STATIC_BOLD_GREEN,
  CCZE_COLOR_STATIC_BOLD_YELLOW,
  CCZE_COLOR_STATIC_BOLD_BLUE,
  CCZE_COLOR_STATIC_BOLD_CYAN,
  CCZE_COLOR_STATIC_BOLD_MAGENTA,
  CCZE_COLOR_STATIC_BOLD_WHITE,
  
  CCZE_COLOR_LAST
} ccze_color_t;

int ccze_color (ccze_color_t idx);
int ccze_color_keyword_lookup (const char *key);

/* Helpers */
ccze_color_t ccze_http_action (const char *method);
void ccze_print_date (const char *date);
char *ccze_strbrk (char *str, char delim);
char *xstrdup (const char *str);

/* Display */
void ccze_addstr (ccze_color_t col, const char *str);
void ccze_newline (void);
void ccze_space (void);
void ccze_wordcolor_process_one (char *word, int slookup);

/* Plugins */
typedef void (*ccze_plugin_startup_t) (void);
typedef void (*ccze_plugin_shutdown_t) (void);
typedef int (*ccze_plugin_handle_t) (const char *str, size_t length,
				     char **rest);

typedef enum
{
  CCZE_PLUGIN_TYPE_FULL,
  CCZE_PLUGIN_TYPE_PARTIAL,
  CCZE_PLUGIN_TYPE_ANY
} ccze_plugin_type_t;

typedef struct
{
  int abi_version;
  void *dlhandle;
  char *name;
  char **argv;
  ccze_plugin_startup_t startup;
  ccze_plugin_shutdown_t shutdown;
  ccze_plugin_handle_t handler;
  ccze_plugin_type_t type;
  char *desc;
} ccze_plugin_t;

char **ccze_plugin_argv_get (const char *name);
const char *ccze_plugin_name_get (void);

#define CCZE_ABI_VERSION 2

#define CCZE_DEFINE_PLUGINS(plugins...) \
char *ccze_plugin_list[] = { plugins, NULL }

#if !defined(BUILTIN)
#define __default_plugin(name) ; \
char ccze_default_plugin[] = # name
#else
#define __default_plugin(name)
#endif

#define CCZE_DEFINE_PLUGIN(name,type,desc) \
ccze_plugin_t ccze_##name##_info = { CCZE_ABI_VERSION, \
				     NULL, \
				     # name, NULL, \
				     ccze_##name##_setup, \
				     ccze_##name##_shutdown, \
				     ccze_##name##_handle, \
				     CCZE_PLUGIN_TYPE_##type, desc } \
__default_plugin (name)

#endif /* !_CCZE_H */


src/mod_DXSpiderPCxx.c
/* -*- mode: c; c-file-style: "gnu" -*-
 * mod_DXSpiderPCxx.c -- DXSpiderPCxx-based on related colorizers for CCZE
 * Copyright (C) 2002, 2003 Gergely Nagy <algernon@bonehunter.rulez.org>
 * Modifications to incorporate DXSpider filtering:
 * Copyright (C) 2023 John Kuras <w7og@yahoo.com>
 *
 * This file is part of ccze.
 *
 * ccze is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * ccze is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 * License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * The modifications to this file made by John Kuras are licensed without
 * hinderance. I.e., do what you want with them.
 */

#include <ccze.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>

#include "ccze-compat.h"

static void ccze_DXSpiderPCxx_setup (void);
static void ccze_DXSpiderPCxx_shutdown (void);
static int ccze_DXSpiderPCxx_handle (const char *str, size_t length, char **rest);

static pcre *reg_DXSpiderPCxx;
static pcre_extra *hints_DXSpiderPCxx;

int isNumber_dx(char *text) {
    int j;
    j = strlen(text);
    while(j--)
    {
        if(isdigit(text[j])) continue;
        if( text[j] == '.' ) continue;
        return 0;
    }
    return 1;
}

static char *
ccze_DXSpiderPCxx_process (const char *str, int *offsets, int match)
{
  char *word = NULL, *toret = NULL;
  char *msg2 = NULL, *num = NULL, *msg = NULL, *hops = NULL, *send = NULL;
  
  pcre_get_substring (str, offsets, match, 1, (const char **)&num);
  pcre_get_substring (str, offsets, match, 2, (const char **)&msg);
  pcre_get_substring (str, offsets, match, 3, (const char **)&hops);
  pcre_get_substring (str, offsets, match, 4, (const char **)&send);
  if ( num[0] == '9' ) {
  	ccze_addstr (CCZE_COLOR_DXSPC9XNUM, "PC");
  	ccze_addstr (CCZE_COLOR_DXSPC9XNUM, num);  
  } else {
  	ccze_addstr (CCZE_COLOR_DXSPCNUM, "PC");
  	ccze_addstr (CCZE_COLOR_DXSPCNUM, num);
  }
  msg2 = xstrdup (msg);
  word = xstrdup (ccze_strbrk (msg2, '^'));
  do {
      if (word != NULL) {
      	  ccze_addstr (CCZE_COLOR_DXSPCDELIM, "^");
		  if ( isNumber_dx(word) ) {
		  	ccze_addstr (CCZE_COLOR_DXSPCMESSAGENUMS, word);
		  } else {
		  	ccze_addstr (CCZE_COLOR_DXSPCMESSAGEWORDS, word);
		  }
		  free(word);
		  word = NULL;
	  } else {
	  	  ccze_addstr (CCZE_COLOR_DXSPCDELIM, "^");
	  }
  } while ((word = xstrdup (ccze_strbrk (NULL, '^'))) != NULL);
  if ( word != NULL ) free(word);
  free (msg2);
  ccze_addstr (CCZE_COLOR_DXSPCDELIM, "^");
  ccze_addstr (CCZE_COLOR_DXSPCHOPS, "H");
  ccze_addstr (CCZE_COLOR_DXSPCHOPS, hops);
  ccze_addstr (CCZE_COLOR_DXSPCDELIM, "^");
  if ( send != NULL ) {
  	  toret = xstrdup (send);
  	  return(toret);
  }
  
  return NULL;
}

static void
ccze_DXSpiderPCxx_setup (void)
{
  const char *error;
  int errptr;
  reg_DXSpiderPCxx = pcre_compile
    ("^PC(\\d{2})\\^(.*?)\\^H(\\d{2})\\^(\\S*)", 0, &error,
     &errptr, NULL);
  hints_DXSpiderPCxx = pcre_study (reg_DXSpiderPCxx, 0, &error);
}

static void
ccze_DXSpiderPCxx_shutdown (void)
{
  free (reg_DXSpiderPCxx);
  free (hints_DXSpiderPCxx);
}

static int
ccze_DXSpiderPCxx_handle (const char *str, size_t length, char **rest)
{
  int match, offsets[99];  
  
  if ( (match = pcre_exec (reg_DXSpiderPCxx, hints_DXSpiderPCxx, str, length, 0, 0, offsets, 99)) >= 0) {
      if (rest) {
		*rest = ccze_DXSpiderPCxx_process (str, offsets, match);
      } else {
		ccze_DXSpiderPCxx_process (str, offsets, match);
      }
      return 1;
  }

  return 0;
}

CCZE_DEFINE_PLUGIN (DXSpiderPCxx, PARTIAL, "Coloriser DXSpider PCxx messages.");


src/mod_DXSpider.c
/* -*- mode: c; c-file-style: "gnu" -*-
 * mod_DXSpiderPCxx.c -- DXSpiderPCxx-based on related colorizers for CCZE
 * Copyright (C) 2002, 2003 Gergely Nagy <algernon@bonehunter.rulez.org>
 * Modifications to incorporate DXSpider filtering:
 * Copyright (C) 2023 John Kuras <w7og@yahoo.com>
 *
 * This file is part of ccze.
 *
 * ccze is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * ccze is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 * License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * The modifications to this file made by John Kuras are licensed without
 * hinderance. I.e., do what you want with them.
 */

#include <ccze.h>
#include <string.h>
#include <stdlib.h>

#include "ccze-compat.h"

static void ccze_DXSpider_setup (void);
static void ccze_DXSpider_shutdown (void);
static int ccze_DXSpider_handle (const char *str, size_t length, char **rest);

static pcre *reg_DXSpider_chan;
static pcre_extra *hints_DXSpider_chan;
static pcre *reg_DXSpider_chanerr;
static pcre_extra *hints_DXSpider_chanerr;
static pcre *reg_DXSpider_progress;
static pcre_extra *hints_DXSpider_progress;

static char *
ccze_DXSpider_process_chan (const char *str, int *offsets, int match)
{
  char *date = NULL, *chan = NULL,  *direction = NULL, *mtype = NULL, *connection = NULL;
  char *protocolMnumber = NULL, *message = NULL;
  char *toret;
  
  pcre_get_substring (str, offsets, match, 1, (const char **)&date);
  pcre_get_substring (str, offsets, match, 2, (const char **)&direction);
  pcre_get_substring (str, offsets, match, 3, (const char **)&mtype);
  pcre_get_substring (str, offsets, match, 4, (const char **)&connection);	/* call */
  pcre_get_substring (str, offsets, match, 5, (const char **)&message);  

  ccze_addstr (CCZE_COLOR_DATE, date);
  ccze_space ();

  ccze_addstr (CCZE_COLOR_DEFAULT, "(");
  ccze_addstr (CCZE_COLOR_DXSCHAN, "chan");
  ccze_addstr (CCZE_COLOR_DEFAULT, ")");
  ccze_space ();
  
  if ( strncmp (mtype,(char*)"I",1) == 0 ) {
  	ccze_addstr (CCZE_COLOR_DXSCHANDIRLEFT, direction);
  	ccze_space ();
  	ccze_addstr (CCZE_COLOR_DXSCHANDIRLEFT, mtype);
  	ccze_space ();
  } else if ( strncmp (mtype,(char*)"D",1) == 0 ) {
  	ccze_addstr (CCZE_COLOR_DXSCHANDIRRIGHT, direction);
  	ccze_space ();
   	ccze_addstr (CCZE_COLOR_DXSCHANDIRRIGHT, mtype);
  	ccze_space ();  
  } else if ( strncmp (mtype,(char*)"X",1) == 0 ) {
  	ccze_addstr (CCZE_COLOR_DXSCHANDIRX, direction);
  	ccze_space ();
  	ccze_addstr (CCZE_COLOR_DXSCHANDIRX, mtype);
  	ccze_space ();  
  } else {
   	ccze_addstr (CCZE_COLOR_DEFAULT, direction);
  	ccze_space ();
  	ccze_addstr (CCZE_COLOR_DEFAULT, mtype);
  	ccze_space ();  
  }
  ccze_addstr (CCZE_COLOR_DXSCHANNAME, connection);
  ccze_space ();
    
  toret = strdup (message);

  free (date);
  free (chan);
  free (direction);
  free (mtype);
  free (connection);
  free (protocolMnumber);
  free (message);
  return toret;
}

static char *
ccze_DXSpider_process_chanerr (const char *str, int *offsets, int match)
{
  char *date = NULL, *chanerr = NULL,  *errortype = NULL;
  char *message = NULL;
  char *toret;
  
  pcre_get_substring (str, offsets, match, 1, (const char **)&date);
  pcre_get_substring (str, offsets, match, 2, (const char **)&errortype);
  pcre_get_substring (str, offsets, match, 3, (const char **)&message);
  

  ccze_addstr (CCZE_COLOR_DATE, date);
  ccze_space ();

  ccze_addstr (CCZE_COLOR_DEFAULT, "(");
  ccze_addstr (CCZE_COLOR_DXSCHANERROR, "chanerr");
  ccze_addstr (CCZE_COLOR_DEFAULT, ")");
  ccze_space ();
  
  ccze_addstr (CCZE_COLOR_DXSCHANERRORTYPE, errortype);
  ccze_addstr (CCZE_COLOR_DEFAULT, ":");
  ccze_space ();


  toret = strdup (message);

  free (date);
  free (chanerr);
  free (errortype);
  free (message);

  return toret;
}

static char *
ccze_DXSpider_process_progress (const char *str, int *offsets, int match)
{
  char *date   = NULL, *errortype = NULL;
  char *dxcall = NULL, *freq      = NULL, *time  = NULL, *spotter = NULL;
  char *node   = NULL, *comment   = NULL, *route = NULL;
  char *toret  = NULL;
  
  pcre_get_substring (str, offsets, match, 1, (const char **)&date);
  pcre_get_substring (str, offsets, match, 2, (const char **)&errortype);// SPOT
  pcre_get_substring (str, offsets, match, 3, (const char **)&dxcall);// call
  pcre_get_substring (str, offsets, match, 4, (const char **)&freq);// freq
  pcre_get_substring (str, offsets, match, 5, (const char **)&time);// time
  pcre_get_substring (str, offsets, match, 6, (const char **)&spotter);// spotter
  pcre_get_substring (str, offsets, match, 7, (const char **)&node);// node
  pcre_get_substring (str, offsets, match, 8, (const char **)&comment);// comment
  pcre_get_substring (str, offsets, match, 9, (const char **)&route);// route
  

  ccze_addstr (CCZE_COLOR_DATE, date);
  ccze_space ();

  ccze_addstr (CCZE_COLOR_DEFAULT, "(");
  ccze_addstr (CCZE_COLOR_DXSPROGRESS, "progress");
  ccze_addstr (CCZE_COLOR_DEFAULT, ")");
  ccze_space ();
  
  ccze_addstr (CCZE_COLOR_DXSPROGRESSTYPE, errortype);
  ccze_addstr (CCZE_COLOR_DXSPROGRESSTYPE, ":");
  ccze_space ();

  ccze_addstr (CCZE_COLOR_DXSDXCALL, dxcall);
  ccze_space ();
  ccze_addstr (CCZE_COLOR_DEFAULT, "on");
  ccze_space ();

  ccze_addstr (CCZE_COLOR_DXSDXFREQUENCY, freq);
  ccze_space ();
  ccze_addstr (CCZE_COLOR_DEFAULT, "@");
  ccze_space ();

  ccze_addstr (CCZE_COLOR_DATE, time);
  ccze_space ();
  ccze_addstr (CCZE_COLOR_DEFAULT, "by");
  ccze_space ();

  ccze_addstr (CCZE_COLOR_DXSDXSPOTTER, spotter);
  ccze_addstr (CCZE_COLOR_DEFAULT, "@");
  
  ccze_addstr (CCZE_COLOR_DXSNODE, node);
  ccze_space ();
  
  ccze_addstr (CCZE_COLOR_DEFAULT, "'");
  ccze_addstr (CCZE_COLOR_DXSDXCOMMENT, comment);
  ccze_addstr (CCZE_COLOR_DEFAULT, "'");
  ccze_space ();
  
  ccze_addstr (CCZE_COLOR_DEFAULT, "route:");
  ccze_space ();
  ccze_addstr (CCZE_COLOR_DXSDXROUTE, route);

  free (date);
  free (dxcall);
  free (errortype);
  free (freq);
  free (time);
  free (spotter);
  free (node);
  free (comment);
  free (route);

  toret = strdup ("");
  return toret;
}

static void
ccze_DXSpider_setup (void)
{
  const char *error;
  int errptr;

  reg_DXSpider_chan = pcre_compile ("^(\\d{1,2}:\\d\\d:\\d\\d)"
			     "\\s\\(chan\\)\\s(->|<-)\\s([IDX])\\s(\\S+)\\s(.*)$", 0, &error,
			     &errptr, NULL);
  hints_DXSpider_chan = pcre_study (reg_DXSpider_chan, 0, &error);
  
  reg_DXSpider_chanerr = pcre_compile ("^(\\d{1,2}:\\d\\d:\\d\\d)"
			     "\\s\\(chanerr\\)\\s(.*):\\s(.*)$", 0, &error,
			     &errptr, NULL);
  hints_DXSpider_chanerr = pcre_study (reg_DXSpider_chanerr, 0, &error);
  
  reg_DXSpider_progress = pcre_compile ("^(\\d{1,2}:\\d\\d:\\d\\d)"
			     "\\s\\(progress\\)\\s(\\S+):\\s(\\S+)\\son\\s(\\S+)"
			     "\\s@\\s(\\S+)\\sby\\s(\\S+)@(\\S+)\\s'(.*?)'\\sroute:\\s(\\S+)$", 0, &error,
			     &errptr, NULL);
  hints_DXSpider_progress = pcre_study (reg_DXSpider_progress, 0, &error);
  
}

static void
ccze_DXSpider_shutdown (void)
{
  free (reg_DXSpider_chan);
  free (hints_DXSpider_chan);
  free (reg_DXSpider_chanerr);
  free (hints_DXSpider_chanerr);
  free (reg_DXSpider_progress);
  free (hints_DXSpider_progress);
}

static int
ccze_DXSpider_handle (const char *str, size_t length, char **rest)
{
  int match, offsets[99];
  
  if ((match = pcre_exec (reg_DXSpider_chan, hints_DXSpider_chan, str, length,
			  0, 0, offsets, 99)) >= 0)
    {
      *rest = ccze_DXSpider_process_chan (str, offsets, match);
      return 1;
    }
    
  if ((match = pcre_exec (reg_DXSpider_chanerr, hints_DXSpider_chanerr, str, length,
			  0, 0, offsets, 99)) >= 0)
    {
      *rest = ccze_DXSpider_process_chanerr (str, offsets, match);
      return 1;
    }
    
  if ((match = pcre_exec (reg_DXSpider_progress, hints_DXSpider_progress, str, length,
			  0, 0, offsets, 99)) >= 0)
    {
      *rest = ccze_DXSpider_process_progress (str, offsets, match);
      return 1;
    }

  return 0;
}

CCZE_DEFINE_PLUGIN (DXSpider, FULL, "Coloriser for DXSpider debug logs.");
